spinalcordtoolbox/reports/qc2.py

```diff
@@ -0,0 +1,352 @@
+        inf_nan_fill(centers[:, 0])
+        inf_nan_fill(centers[:, 1])
+
+        # Generate the first QC report image
+        img = equalize_histogram(mosaic(img_input, centers))
+
+        # For QC reports, axial mosaics will often have smaller height than width
+        # (e.g. WxH = 20x3 slice images). So, we want to reduce the fig height to match this.
+        # `size_fig` is in inches. So, dpi=300 --> 1500px, dpi=100 --> 500px, etc.
+        size_fig = [5, 5 * img.shape[0] / img.shape[1]]
+
+        fig = Figure()
+        fig.set_size_inches(*size_fig, forward=True)
+        FigureCanvas(fig)
+        ax = fig.add_axes((0, 0, 1, 1))
+        ax.imshow(img, cmap='gray', interpolation='none', aspect=1.0)
+        add_orientation_labels(ax)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+        img_path = str(path_background_img)
+        logger.debug('Save image %s', img_path)
+        fig.savefig(img_path, format='png', transparent=True, dpi=300)
+
+        # Generate the second QC report image
+        img = equalize_histogram(mosaic(img_output, centers))
+        fig = Figure()
+        fig.set_size_inches(*size_fig, forward=True)
+        FigureCanvas(fig)
+        ax = fig.add_axes((0, 0, 1, 1), label='0')
+        ax.imshow(img, cmap='gray', interpolation='none', aspect=1.0)
+        add_orientation_labels(ax)
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+        img_path = str(path_overlay_img)
+        logger.debug('Save image %s', img_path)
+        fig.savefig(img_path, format='png', transparent=True, dpi=300)
+
+
+def inf_nan_fill(A: np.ndarray):
+    """
+    Interpolate inf and NaN values with neighboring values in a 1D array, in-place.
+    If only inf and NaNs, fills the array with zeros.
+    """
+    valid = np.isfinite(A)
+    invalid = ~valid
+    if np.all(invalid):
+        A.fill(0)
+    elif np.any(invalid):
+        A[invalid] = np.interp(
+            np.nonzero(invalid)[0],
+            np.nonzero(valid)[0],
+            A[valid])
+
+
+def mosaic(img: Image, centers: np.ndarray, radius: tuple[int, int] = (15, 15)):
+    """
+    Arrange the slices of `img` into a grid of images.
+
+    Each slice is centered at the approximate coordinates given in `centers`,
+    and cropped to `radius` pixels in each direction (horizontal, vertical).
+
+    If `img` has N slices, then `centers` should have shape (N, 2).
+    """
+    # Fit as many slices as possible in each row of 600 pixels
+    num_col = math.floor(600 / (2*radius[0]))
+
+    # Center and crop each axial slice
+    cropped = []
+    for center, slice in zip(centers.astype(int), img.data):
+        # Add a margin before cropping, in case the center is too close to the edge
+        cropped.append(np.pad(slice, radius)[
+            center[0]:center[0] + 2*radius[0],
+            center[1]:center[1] + 2*radius[1],
+        ])
```

Comment:
This is actually a change in behaviour: the existing code returns a smaller patch instead of adding padding, when the center is too close to the edge. But this interacts poorly with the existing mosaic code, and the too-small patch ends up in the wrong location (which is a bug).

Reply:
For posterity: It was indeed a bug! You predicted it here months before it was reported: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4386